nx/lib/nx/defn.ex

```diff
@@ -654,12 +654,67 @@ defmodule Nx.Defn do
   all local function calls within `defn`.
   """
   defmacro defnp(call, do: block) do
-    define(:defp, call, block, __CALLER__)
+    define_defn(:defp, call, block, __CALLER__)
+  end
+
+  @doc """
+  Defines a transform that executes the given `fun` with `arg`
+  when building `defn` expressions.
+
+  ## Example
+
+  Take the following defn expression:
+
+      defn tanh_power(a, b) do
+        Nx.tanh(a) + Nx.power(b, 2)
+      end
+
+  Let's see a trivial example, which is to use `IO.inspect/1` to
+  print a tensor expression at definition time:
+
+      defn tanh_power(a, b) do
+        Nx.tanh(a) + Nx.power(b, 2) |> my_inspect()
+      end
+
+      deftransformp my_inspect(expr), do: IO.inspect(expr)
+
+  Or:
+
+      defn tanh_power(a, b) do
+        res = Nx.tanh(a) + Nx.power(b, 2)
+        my_inspect(res)
+        res
+      end
+
+  When invoked in both cases, it will print the expression being built
+  by `defn`:
+
+      #Nx.Defn.Expr<
+        parameter a
+        parameter c
+        b = tanh [ a ] ()
+        d = power [ c, 2 ] ()
+        e = add [ b, d ] ()
+      >
+
+  Although, for convenience, you might use `inspect_expr/2` instead.
+  """
+  defmacro deftransform(call, body \\ []) do
+    block = body[:do]
+    define_transform(:def, call, block, __CALLER__)
+  end
+
+  @doc """
+  Private function version for `deftransform`
+  """
+  defmacro deftransformp(call, body \\ []) do
+    block = body[:do]
+    define_transform(:defp, call, block, __CALLER__)
```

Comment:
style question: the current or the following?

```suggestion
    define_transform(:defp, call, body[:do], __CALLER__)
```

Reply:
Or even a separate /1 clause with it's own docs (what would even go there, actually?)

Reply:
I think you should define two separate clauses. Your code as-is  would accept this:

    deftransform foo(a, b), doe: a + b

So make the arity 1 their own functions and explicitly match on `deftransform (call, do: block)` for arity 2. :)

Reply:
What's the issue with accepting the definition like that?

Reply:
Oooh I see, the typo was intentional :)